File: minecraftmod/api/src/main/java/cn/li/acapi/wireless/WirelessImc.java
package cn.li.acapi.wireless;

/**
 * Inter-Mod Communication (IMC) constants for the wireless energy system.
 *
 * <p>External mods observe wireless topology changes by sending an IMC message
 * during {@code InterModEnqueueEvent} (Forge):
 *
 * <pre>{@code
 * InterModComms.sendTo(WirelessImc.MOD_ID, WirelessImc.REGISTER_NETWORK_HANDLER,
 *     () -> (java.util.function.Consumer<java.util.Map<?, ?>>) event -> {
 *         // inspect the event map, see key contract below
 *     });
 * }</pre>
 *
 * <p><b>Payload:</b> the message payload must be a
 * {@code java.util.function.Consumer<java.util.Map<?, ?>>} (Clojure callers may
 * pass an {@code IFn} instead). Handlers are invoked on the server thread after
 * the corresponding topology change. A handler that throws is logged and removed.
 *
 * <p><b>Event map contract</b> — keys are Clojure keywords
 * ({@code clojure.lang.Keyword.intern("action")} etc.), values as listed:
 *
 * <ul>
 *   <li>Network events ({@link #REGISTER_NETWORK_HANDLER}):
 *     <ul>
 *       <li>{@code :kind} — {@code :topology/network}</li>
 *       <li>{@code :action} — {@code :created} | {@code :destroyed}</li>
 *       <li>{@code :ssid} — {@code String}</li>
 *       <li>{@code :matrix} — {@link IWirelessMatrix}</li>
 *     </ul>
 *   </li>
 *   <li>Node events ({@link #REGISTER_NODE_HANDLER}):
 *     <ul>
 *       <li>{@code :kind} — {@code :topology/node}</li>
 *       <li>{@code :action} — {@code :connected} | {@code :disconnected} |
 *           {@code :generator-linked} | {@code :generator-unlinked} |
 *           {@code :receiver-linked} | {@code :receiver-unlinked}</li>
 *       <li>{@code :node} — {@link IWirelessNode}</li>
 *       <li>{@code :matrix} — {@link IWirelessMatrix}, present on
 *           {@code :connected} / {@code :disconnected}</li>
 *       <li>{@code :generator} — {@link IWirelessGenerator}, present on
 *           generator link events</li>
 *       <li>{@code :receiver} — {@link IWirelessReceiver}, present on
 *           receiver link events</li>
 *     </ul>
 *   </li>
 * </ul>
 */
public final class WirelessImc {

    /** The mod-id of this mod; use as target id in platform messaging. */
    public static final String MOD_ID = "academycraft";

    /** IMC key: register a network topology event consumer. */
    public static final String REGISTER_NETWORK_HANDLER = "register_wireless_network_handler";

    /** IMC key: register a node topology event consumer. */
    public static final String REGISTER_NODE_HANDLER = "register_wireless_node_handler";

    private WirelessImc() {}
}

File: minecraftmod/forge-1.20.1/src/main/clojure/cn/li/forge1201/wireless_imc.clj
(ns cn.li.forge1201.wireless-imc
  "IMC (Inter-Mod Communication) support for the wireless energy system.

  External mods register handlers during InterModEnqueueEvent by sending IMC
  messages to this mod using the keys defined in WirelessImc. Handlers are
  invoked after the corresponding Forge event fires, so they receive the same
  information as a regular EventBus subscriber but via callback.

  Handler isolation: a handler that throws is logged at DEBUG and removed so
  it cannot disrupt subsequent ticks or other handlers."
  (:require [cn.li.mcmod.util.log :as log])
  (:import [cn.li.acapi.wireless WirelessImc$NetworkEventHandler WirelessImc$NodeEventHandler]
           [cn.li.acapi.wireless.event WirelessNetworkEvent$NetworkCreated
                                       WirelessNetworkEvent$NetworkDestroyed
                                       WirelessNetworkEvent$NodeConnected
                                       WirelessNetworkEvent$NodeDisconnected
                                       WirelessNetworkEvent$GeneratorLinked
                                       WirelessNetworkEvent$ReceiverLinked]))

;; ============================================================================
;; Handler registries
;; ============================================================================

(defonce ^:private network-handlers (atom []))
(defonce ^:private node-handlers    (atom []))

(defn register-network-handler!
  "Register an IMC network event handler. Called from InterModProcessEvent."
  [^WirelessImc$NetworkEventHandler handler]
  (swap! network-handlers conj handler))

(defn register-node-handler!
  "Register an IMC node event handler. Called from InterModProcessEvent."
  [^WirelessImc$NodeEventHandler handler]
  (swap! node-handlers conj handler))

;; ============================================================================
;; Dispatch helpers
;; ============================================================================

(defn- invoke-safe
  "Call f, return nil. If it throws, log and return ::remove-handler."
  [f]
  (try (f) nil
       (catch Exception e
         (log/debug "IMC wireless handler threw exception, removing it:" (ex-message e))
         ::remove-handler)))

(defn- dispatch-network! [type ssid matrix]
  (let [bad (keep (fn [h]
                    (when (= ::remove-handler
                              (invoke-safe #(.onNetworkEvent ^WirelessImc$NetworkEventHandler h
                                                            type ssid matrix)))
                      h))
                  @network-handlers)]
    (when (seq bad)
      (swap! network-handlers #(remove (set bad) %)))))

(defn- dispatch-node! [type node]
  (let [bad (keep (fn [h]
                    (when (= ::remove-handler
                              (invoke-safe #(.onNodeEvent ^WirelessImc$NodeEventHandler h type node)))
                      h))
                  @node-handlers)]
    (when (seq bad)
      (swap! node-handlers #(remove (set bad) %)))))

;; ============================================================================
;; Event dispatch – dispatches wireless payload events to IMC handlers
;; ============================================================================

(defn- on-wireless-event [event]
  (cond
    (instance? WirelessNetworkEvent$NetworkCreated event)
    (let [^WirelessNetworkEvent$NetworkCreated e event]
      (dispatch-network! "created" (.getSsid e) (.getMatrix e)))

    (instance? WirelessNetworkEvent$NetworkDestroyed event)
    (let [^WirelessNetworkEvent$NetworkDestroyed e event]
      (dispatch-network! "destroyed" (.getSsid e) (.getMatrix e)))

    (instance? WirelessNetworkEvent$NodeConnected event)
    (let [^WirelessNetworkEvent$NodeConnected e event]
      (dispatch-node! "connected" (.getNode e)))

    (instance? WirelessNetworkEvent$NodeDisconnected event)
    (let [^WirelessNetworkEvent$NodeDisconnected e event]
      (dispatch-node! "disconnected" (.getNode e)))

    (instance? WirelessNetworkEvent$GeneratorLinked event)
    (let [^WirelessNetworkEvent$GeneratorLinked e event]
      (dispatch-node! "generator_linked" (.getNode e)))

    (instance? WirelessNetworkEvent$ReceiverLinked event)
    (let [^WirelessNetworkEvent$ReceiverLinked e event]
      (dispatch-node! "receiver_linked" (.getNode e)))

    :else nil))

(defn dispatch-event!
  "Dispatch a wireless runtime payload event to registered IMC handlers."
  [event]
  (on-wireless-event event)
  nil)

(defn init!
  "Initialize wireless IMC bridge. Dispatch is bound through platform-events."
  []
  (log/info "Wireless IMC dispatcher ready (direct payload dispatch mode)"))
